src/targets/gpu/include/migraphx/gpu/reduce_op.hpp

```diff
@@ -0,0 +1,77 @@
+#ifndef MIGRAPHX_GUARD_RTGLIB_REDUCE_OP_HPP
+#define MIGRAPHX_GUARD_RTGLIB_REDUCE_OP_HPP
+
+#include <migraphx/gpu/hip.hpp>
+#include <migraphx/gpu/context.hpp>
+#include <migraphx/shape.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/type_name.hpp>
+#include <utility>
+#include <iostream>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+
+struct context;
+
+template <class Derived, class Op, void (*F)(hipStream_t, const argument&, const argument&)>
+struct reduce_op
+{
+    Op op;
+
+    template <class Self, class T>
+    static auto reflect(Self& self, T f)
+    {
+        return migraphx::reflect(self.op, f);
+    }
+
+    std::string name() const
+    {
+        const std::string& name = get_type_name<Derived>();
+        // search the namespace gpu (::gpu::)
+        auto pos_ns = name.find("::gpu::");
```

Comment:
can we use the utility function contains(name, "::gpu::") for this?

Reply:
We may not be able to since we need to pos info returned from the find function, while contain() does not return that. 

As for the purpose, the name returned from get_type_name() contains other info like namespace that we donot want. These code is to retrieve the actually operator name we want.

Reply:
can you clarify this behavior of removing namespace info in the code?

Reply:
for example, the name returned from get_type_name<Derived>() is "migraphx::version_1::gpu::hip_reduce_sum", this function is to extract the part "reduce_mean", then with the prefix "gpu::", we get the operator name "gpu::reduce_mean"

Reply:
Sorry... I meant clarify to the reader by adding comments

Reply:
OK, will add some comments for that.